libdnf/sack/query.cpp

```diff
@@ -1022,32 +1038,73 @@ Query::Impl::filterNevraStrict(int cmpType, const char **matches)
     Map nevraResult;
     map_init(&nevraResult, pool->nsolvables);
 
-    if (compareSet.size() > 1) {
-        std::sort(compareSet.begin(), compareSet.end(), nevraIDSorter);
+    //  if cmpType == HY_EQ or cmpType == (HY_EQ | HY_NOT) -> performance optimization
+    if (createNewEVR == 0) {
+        if (compareSet.size() > 1) {
+            std::sort(compareSet.begin(), compareSet.end(), nevraIDSorter);
 
-        Id id = -1;
-        while (true) {
-            id = result->next(id);
-            if (id == -1)
-                break;
-            Solvable* s = pool_id2solvable(pool, id);
-            auto low = std::lower_bound(compareSet.begin(), compareSet.end(), *s,
-                                        nevraCompareLowerSolvable);
-            if (low != compareSet.end() && low->name == s->name && low->evr == s->evr &&
-                low->arch == s->arch) {
-                MAPSET(&nevraResult, id);
+            Id id = -1;
+            while (true) {
+                id = result->next(id);
+                if (id == -1)
+                    break;
+                Solvable* s = pool_id2solvable(pool, id);
+                auto low = std::lower_bound(compareSet.begin(), compareSet.end(), *s,
+                                            nevraCompareLowerSolvable);
+                if (low != compareSet.end() && low->name == s->name && low->evr == s->evr &&
+                    low->arch == s->arch) {
+                    MAPSET(&nevraResult, id);
+                }
+            }
+        } else {
+            auto nevraId = compareSet[0];
+            Id id = -1;
+            while (true) {
+                id = result->next(id);
+                if (id == -1)
+                    break;
+                Solvable* s = pool_id2solvable(pool, id);
+                if (nevraId.name == s->name && nevraId.evr == s->evr && nevraId.arch == s->arch) {
+                    MAPSET(&nevraResult, id);
+                }
             }
         }
     } else {
-        auto nevraId = compareSet[0];
-        Id id = -1;
-        while (true) {
-            id = result->next(id);
-            if (id == -1)
-                break;
-            Solvable* s = pool_id2solvable(pool, id);
-            if (nevraId.name == s->name && nevraId.evr == s->evr && nevraId.arch == s->arch) {
-                MAPSET(&nevraResult, id);
+        if (compareSet.size() > 1) {
+            std::sort(compareSet.begin(), compareSet.end(), nevraIDSorter);
+
+            Id id = -1;
+            while (true) {
+                id = result->next(id);
+                if (id == -1)
+                    break;
+                Solvable* s = pool_id2solvable(pool, id);
+                auto low = std::lower_bound(compareSet.begin(), compareSet.end(), *s,
+                                            nameArchCompareLowerSolvable);
+                if (low != compareSet.end() && low->name == s->name && low->arch == s->arch) {
```

Comment:
It will only work if the `matches` array contains NEVRAs with a unique combination of name and architecture, right? Only the first matching name-arch item from `compareSet` is used.
Probably a filter that contains multiple items with the same name and architecture does not make sense with `HY_LT`/`HY_GT`. And your simplification is not a problem.